(WIP) ES6 Source

kevinbeaty opened the pull request:
Resolves #33

kevinbeaty commented:
@tgriesser Nice. Thanks.  I'll probably manually cherry pick things from your translation.   I've started moving to ES6 subclasses in the PR for transducers and I now prefer that approach for internal implementations over the use of `transducer`.  I do think your original point stands as it makes the code more difficult than necessary to follow and likely will negatively affect performance.  With classes, my objections to code duplication are greatly diminished. 

I'm also planning on flattening modules to a few core files to resolve #34 .  I still want to maintain modularity at that level as I'd like to be able to use, e.g. `transduce/string` without requiring `transducer` or `core` in a bundle if I'm not using it.  You have convinced me that function level file splitting is overkill though, so hopefully flattening one level will make things easier to review. I  already have a solution for this in the PR, just a matter of migrating the other files and possibly reorganizing slightly. 

I'll will definitely solicit your comments once I have things in a reviewable state.  Would appreciate any feedback when that time comes.

I did notice you were using webpack instead of browserify.  What are the benefits from your perspective?  Genuinely curious, as I have never spent the time to review the differences...

Thanks, as always, for your feedback.

kevinbeaty commented:
Resolves #31
Resolves #34

tgriesser commented:
Just FYI you no longer need to write `use strict`, es6 is strict by default and babel adds it for you automatically

kevinbeaty commented:
> Just FYI you no longer need to write use strict, es6 is strict by default and babel adds it for you automatically

Thanks.  JSHint is complaining when I remove that. I'll have to figure a way to turn that off.

I think that this is ready to merge now, outside any other suggestions you have....

kevinbeaty commented:
> Thanks. JSHint is complaining when I remove that. I'll have to figure a way to turn that off.

Well, I made that problem much harder than it should have been.  Just needed to remove a line from my rc file :)

tgriesser commented:
> I think that this is ready to merge now, outside any other suggestions you have....

Cool, I'm playing around with it a bit... I think i'm going to keep on my own local fork of it until i'm comfortable with some things.

One thought I had was that `init`, `step`, `result` should be renamed to `xfInit`, `xfStep`, `xfResult`, the former is especially confusing given the recent api change

tgriesser commented:
Also you can take out the `_keys` shim... there's a bunch of places elsewhere that things will break if there's not an es5 shim so we don't need to polyfill that ourselves.

tgriesser commented:
Also, I'm thinking the function iterable implementation should be something more like:

```js
export function iterable(val){
  var it
  if(value[symIter] !== void 0){
    it = value
  } else if(isArray(val) || isString(val)){
    it = new ArrayIterable(val)
  } else if(isFunction(val)){
    it = new FunctionIterable(() => {done: false, value: val()})
  } else if(isFunction(value.next)){
    it = new FunctionIterable(() => val.next())
  } else {
    it = new ObjectIterable(value)
  }
  return it
}

class FunctionIterable {
  constructor(fn){
    this.fn = fn
  }
  [symIter](){
    return {
      next: () => this.fn()
    }
  }
}
```

Right? Since you'll want the actual value from the `val.next` rather than doing 
```js
{done: false, value: val.next()}
```

tgriesser commented:
Oh, one other thing I was going to ask about was the setting of everything to local var's at the top of a function:

For example [here](https://github.com/transduce/transduce/blob/4a00a8a46021c417d7dd8b8206f9446718618bb2/src/core.js#L128-L130) it's not exactly clear why these are set to local vars... other than to shave a byte or two on minification I'd assume. 

In other cases like [here](https://github.com/transduce/transduce/blob/4a00a8a46021c417d7dd8b8206f9446718618bb2/src/transducers.js#L146-L149) the assignment actually matters because you're capturing and reassigning the values, but in other cases I think it just makes things slightly more difficult to read, since you're creating extra vars for no reason. 

Might be more of a matter of opinion though...

kevinbeaty commented:
> One thought I had was that init, step, result should be renamed to xfInit, xfStep, xfResult, the former is especially confusing given the recent api change

Good idea.  I've changed it as suggested.

> Also you can take out the _keys shim.

Removed

> Also, I'm thinking the function iterable implementation should be something more like:

Thanks.  Fixed and added a test with an object with `next`

> Oh, one other thing I was going to ask about was the setting of everything to local var's at the top of a function:

This is just a habit I picked up long ago. All variable declarations at the top.  Really no reason I do this, and you could convince me to change.  I'll probably leave this as is for now though.

tgriesser commented:
> All variable declarations at the top.

All variable declarations at the top is fine, just aliasing local variables when otherwise unnecessary is more what I'm asking about in that comment.

kevinbeaty commented:
Ah.  Gotcha.  I'll do a once over to clean some of that up.

Now that we changed to `xfStep` for stepping on wrapped transformer, I'm thinking of adding `step` as an alias to `'@@transducer/step'` in the base `Transducer` class.  That way we wouldn't have to import (or define) the symbols everywhere, just override `step`.  What do you think?

kevinbeaty commented:
> just aliasing local variables when otherwise unnecessary

I've cleaned it up.

> I'm thinking of adding step as an alias to '@@transducer/step' in the base Transducer class.

Eh. I don't think this is worth it anymore

I think I'm ready to merge again.

tgriesser commented on base.js:
```diff
@@ -1,4 +1 @@
-var core = require('./core'),
-    merge = core.util.objectMerge
-module.exports = merge(merge({}, core), require('./transducers'))
+module.exports = require('./lib/base')
```
If you wanted, I believe `export * from './lib/base'` would have the same effect

tgriesser commented:
Once you merge it can you wait before cutting a new version... I had a few things I wanted to PR on with the es6 and see what you think

kevinbeaty commented:
> If you wanted, I believe export * from './lib/base' would have the same effect

This files are here only to support `require('transduce/string')` etc. in Node.js/browserify.  The current build only runs babel from `src/*.js -> lib/*.js` and does not touch these top level files so I think they need to stay using Node.js `require` here.  I'd like to come up a way to get rid of these files altogether and still allow the require relative to transduce but I think this is an OK stopgap for now.

> Once you merge it can you wait before cutting a new version... I had a few things I wanted to PR on with the es6 and see what you think

I'll merge now.  PR at will.
